geometry/volume.hpp

```diff
@@ -2004,100 +2027,223 @@ void ScalarField_t<Value_t, Container_t>::isoFromTetrahedron(
 
 }
 
-template<typename Value_t, class Container_t>
+/*
+ *
+ * |k
+ * |  /j
+ * | /
+ * |/      i
+ * +--------
+ *
+ * local vertex and edge ids
+ *
+ *     v7------e6-------v6
+ *    /|               /|
+ *   e7|             e5 |
+ *  / e11            /  e10
+ * v4-------e4-----v5   |
+ * |   |           |    |
+ * e8  |           e9   |
+ * |   v3------e2--|---v2
+ * |  /            |   /
+ * | e3            |  e1
+ * |/              | /
+ * v0------e0------v1
+ *
+ */
+template <typename Value_t, class Container_t>
 void ScalarField_t<Value_t, Container_t>::isoFromCube(
-        std::vector<typename VolumeBase_t::FPosition_s> & retval
-        , const typename VolumeBase_t::FPosition_s * vertices
-        , const Value_t * values
-        , const Value_t & threshold, const SurfaceOrientation_t orientation)
-const
+    std::vector<std::pair<typename VolumeBase_t::FPosition_s, std::size_t>>&
+        retval,
+    const typename VolumeBase_t::FPosition_s* vertices,
+    const Value_t* values,
+    const Value_t& threshold,
+    const SurfaceOrientation_t orientation,
+    std::size_t i,
+    std::size_t j,
+    std::size_t k) const
 {
     typedef typename VolumeBase_t::FPosition_s FPosition_s;
 
-    int cubeIndex;
-    FPosition_s vertexList[12];
-
-    cubeIndex = 0;
-    if(orientation == TO_MIN){
-        if (values[0] < threshold) cubeIndex |= 1;
-        if (values[1] < threshold) cubeIndex |= 2;
-        if (values[2] < threshold) cubeIndex |= 4;
-        if (values[3] < threshold) cubeIndex |= 8;
-        if (values[4] < threshold) cubeIndex |= 16;
-        if (values[5] < threshold) cubeIndex |= 32;
-        if (values[6] < threshold) cubeIndex |= 64;
-        if (values[7] < threshold) cubeIndex |= 128;
-    }else{
-        if (values[0] > threshold) cubeIndex |= 1;
-        if (values[1] > threshold) cubeIndex |= 2;
-        if (values[2] > threshold) cubeIndex |= 4;
-        if (values[3] > threshold) cubeIndex |= 8;
-        if (values[4] > threshold) cubeIndex |= 16;
-        if (values[5] > threshold) cubeIndex |= 32;
-        if (values[6] > threshold) cubeIndex |= 64;
-        if (values[7] > threshold) cubeIndex |= 128;
-    }
-
-    if (marchingcubes::edgeTable[cubeIndex] == 0)
-        return;
+    std::pair<FPosition_s, std::size_t> vertexList[12];
+
+    auto edgeId = [this, i, j, k](int localEdge) -> std::size_t {
+        std::size_t si = this->container_.sizeX() + 1;
+        std::size_t sj = this->container_.sizeY() + 1;
+
+        switch (localEdge)
+        {
+        case 0:
+            return 3 * i + 3 * j * (si + 1) + 3 * k * (si + 1) * (sj + 1);
```

Comment:
Maybe a short comment what's the idea behind the indexing? The formulas could be a little more readable, maybe omit the spaces around "`*`" to make the terms more readable? Also `si`, `sj` are always used as `(si+1)`, `(sj+1)`, maybe it would make the formulas shorter to include the `+1` in the values up-front? (But maybe there's a reason for the +1 here, this could be explained). It would also be nice to document if the indexing is contiguous or if some of the indices are unused (it's not immediately apparent). Could perhaps some common sub-expression be factored out, like the cube index `3*(i + j*(si+1) + k*(si+1)*(sj+1))`?

Reply:
I think `(si + 1)` and `(si + 1)*(sj + 1)` could be called something ("stride"?) to simplify the expressions...

Reply:
Redesigned and simplified. Hopefully this is more readable. Algorithm descriptiom added.